cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -840,39 +836,53 @@ void LocalVersionedEngine::delete_trees_responsibly(
     not_to_delete.clear();
     log::version().debug("Forbidden: {} total of data keys", data_keys_not_to_be_deleted.size());
-
     RemoveOpts remove_opts;
     remove_opts.ignores_missing_key_ = true;
-
-    std::vector<entity::VariantKey> vks;
-    // Delete any associated column stats keys first
+    folly::Future<folly::Unit> write_version_fut;
     if (!dry_run) {
+        std::vector<entity::VariantKey> vks_column_stats;
+        // Delete any associated column stats keys first
         std::transform(keys_to_delete->begin(),
                        keys_to_delete->end(),
-                       std::back_inserter(vks),
+                       std::back_inserter(vks_column_stats),
                        [](const IndexTypeKey& index_key) {
             return index_key_to_column_stats_key(index_key);
         });
-        store()->remove_keys(vks, remove_opts).get();
-    }
 
-    vks.clear();
-    if (!dry_run) {
-        std::copy(keys_to_delete->begin(), keys_to_delete->end(), std::back_inserter(vks));
-        store()->remove_keys(vks, remove_opts).get();
-    }
-
-    vks.clear();
-    std::copy_if(std::make_move_iterator(data_keys_to_be_deleted.begin()),
-                 std::make_move_iterator(data_keys_to_be_deleted.end()),
-                 std::back_inserter(vks),
-                 [&](const auto& k) {return !data_keys_not_to_be_deleted.count(k);});
-    log::version().debug("Index keys deleted. Proceeding with {} number of data keys", vks.size());
-    if (!dry_run) {
-        store()->remove_keys(vks, remove_opts).get();
+        std::vector<entity::VariantKey> vks_to_delete;
+        std::copy(std::make_move_iterator(keys_to_delete->begin()),
+                    std::make_move_iterator(keys_to_delete->end()),
+                    std::back_inserter(vks_to_delete));
+
+        write_version_fut = store()->remove_keys(std::move(vks_column_stats), remove_opts)
+        .thenValue([this, vks_to_delete = std::move(vks_to_delete), remove_opts](auto&& ) mutable {
+            return store()->remove_keys(std::move(vks_to_delete), remove_opts);
+        })
+        .thenValue([](auto&&){
+            return folly::Unit();
+        });
     }
+    return std::move(write_version_fut)
+    .thenValue([this, data_keys_to_be_deleted = std::move(data_keys_to_be_deleted), data_keys_not_to_be_deleted = std::move(data_keys_not_to_be_deleted), remove_opts, dry_run](auto&& ){
+        std::vector<entity::VariantKey> vks_data_to_delete;
+        std::copy_if(std::make_move_iterator(data_keys_to_be_deleted.begin()),
+                    std::make_move_iterator(data_keys_to_be_deleted.end()),
+                    std::back_inserter(vks_data_to_delete),
+                    [&](const auto& k) {return !data_keys_not_to_be_deleted.count(k);});
+
+        log::version().debug("Index keys deleted. Proceeding with {} number of data keys", vks_data_to_delete.size());
+        folly::Future<folly::Unit> ret_fut;
+        if (!dry_run) {
```

Comment:
This can be removed once the return is one long `thenValue` chain starting on line 857

Reply:
As agreed, in order to leave the code cleaner, we will increase the memory footprint by constructing the three vectors in advance